gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/iceberg/IcebergDatasetTest.java

```diff
@@ -307,105 +329,137 @@ protected IcebergTable validateGetFilePathsGivenDestState(
+    public TrickIcebergDataset(String db, String table, IcebergTable icebergTbl, Properties properties,
+        FileSystem sourceFs) {
       super(db, table, icebergTbl, properties, sourceFs);
     }
 
     @Override // as the `static` is not mock-able
-    protected FileSystem getSourceFileSystemFromFileStatus(FileStatus fileStatus, Configuration hadoopConfig) throws IOException {
+    protected FileSystem getSourceFileSystemFromFileStatus(FileStatus fileStatus, Configuration hadoopConfig)
+        throws IOException {
       return this.sourceFs;
     }
-  };
+  }
 
+  ;
 
   /** Builds a {@link FileSystem} mock */
   protected static class MockFileSystemBuilder {
     private final URI fsURI;
     /** when not `.isPresent()`, all paths exist; when `.get().isEmpty()`, none exist; else only those indicated do */
-    private final Optional<Set<Path>> optPaths;
+    private final Optional<Map<Path, FileStatus>> optPathsWithFileStatuses;
 
     public MockFileSystemBuilder(URI fsURI) {
       this(fsURI, false);
     }
 
     public MockFileSystemBuilder(URI fsURI, boolean shouldRepresentEveryPath) {
       this.fsURI = fsURI;
-      this.optPaths = shouldRepresentEveryPath ? Optional.empty() : Optional.of(Sets.newHashSet());
+      this.optPathsWithFileStatuses = shouldRepresentEveryPath ? Optional.empty() : Optional.of(Maps.newHashMap());
     }
 
-    public Optional<Set<Path>> getPaths() {
-      return this.optPaths.map(Sets::newHashSet); // copy before returning
+    public void addPaths(List<String> pathStrings) {
+      addPathsAndFileStatuses(pathStrings, false);
     }
 
-    public void addPaths(List<String> pathStrings) {
+    public void addPathsAndFileStatuses(List<String> pathStrings, boolean shouldCreateFileStatusWithPermissions) {
       for (String pathString : pathStrings) {
-        addPath(pathString);
+        addPathAndFileStatus(pathString, shouldCreateFileStatusWithPermissions);
       }
     }
 
-    public void addPath(String pathString) {
-      addPath(new Path(pathString));
+    public void addPathAndFileStatus(String pathString, boolean shouldCreateFileStatusWithPermissions) {
+      addPathAndFileStatus(new Path(pathString), shouldCreateFileStatusWithPermissions);
     }
 
-    public void addPath(Path path) {
-      if (!this.optPaths.isPresent()) {
+    public void addPathAndFileStatus(Path path, boolean shouldCreateFileStatusWithPermissions) {
+      if (!this.optPathsWithFileStatuses.isPresent()) {
         throw new IllegalStateException("unable to add paths when constructed with `shouldRepresentEveryPath == true`");
       }
-      if (this.optPaths.get().add(path) && !path.isRoot()) { // recursively add ancestors of a previously unknown path
-        addPath(path.getParent());
+      FileStatus fileStatus = shouldCreateFileStatusWithPermissions ? createFileStatus(path, "test_owner", "test_group", new FsPermission(FsAction.WRITE_EXECUTE, FsAction.READ_EXECUTE, FsAction.NONE)) : null;
```

Comment:
hard-coding behind-the-scenes makes the mock brittle--and code harder to follow.  let's keep things simple w/ the code using the mock choosing the filestatus to supply for which path.